src/coreclr/jit/valuenum.cpp

```diff
@@ -9329,143 +9342,169 @@ void Compiler::fgValueNumberSimd(GenTreeSIMD* tree)
 #endif // FEATURE_SIMD
 
 #ifdef FEATURE_HW_INTRINSICS
-// Does value-numbering for a GT_HWINTRINSIC node
 void Compiler::fgValueNumberHWIntrinsic(GenTreeHWIntrinsic* tree)
 {
-    // For safety/correctness we must mutate the global heap valuenumber
-    // for any HW intrinsic that performs a memory store operation
-    if (tree->OperIsMemoryStore())
-    {
-        fgMutateGcHeap(tree DEBUGARG("HWIntrinsic - MemoryStore"));
-    }
+    NamedIntrinsic intrinsicId   = tree->GetHWIntrinsicId();
+    GenTree*       addr          = nullptr;
+    const bool     isMemoryLoad  = tree->OperIsMemoryLoad(&addr);
+    const bool     isMemoryStore = tree->OperIsMemoryStore(&addr);
```

Comment:
This seems like bad/confusing behavior as its relying on `OperIsMemoryStore` not mutating `addr` on `false`.

Reply:
I understand that's fine for how it currently is, but it would be good if we made this more explicit and ensure it doesn't look like a potential bug if you're just glancing through the code.

Reply:
Heh, it is actually deliberate. I agree it is not great though, will change it to the original of ``isMemoryStore = !isMemoryLoad && tree->OperIsMemoryStore(&addr);``.